### DISCUSSION TITLE
Restrict child applications to use certain projects

### DISCUSSION THREAD

**Discussion by paralika** [argoproj/argo-cd#8251] [Q&A] [answered]
Can someone please help me to understand if I can restrict child applications to use certain projects only? 

I am trying to use app of apps patterns in my multi-tenant environment. I am controlling root applications as an admin and allowing developers to create child apps using repositories. But I did not find any way to restrict them to use certain projects only as it is important in multi-tenant environment.

**Comment by andrewm-aero** [comment, marked as answer]
As far as I know, this has been a known problem and deficiency for years: https://github.com/argoproj/argo-cd/issues/2785 .

The best you can hope for is per-tenant argocd installs. Those installs need to
* Have their own namespace for argocd resources, per-tenant, and separate from their "application" namespaces
* Have its own service account in that namespace for argocd, which can
    * Create applications and projects in the same namespace
    * Create the types of resources in the namespaces and clusters that they would be limited to via projects in a single-install setup

In particular, if they are allowed to create RBAC resources (serviceaccount, role, rolebinding, etc) in their per-tenant argocd namespace, or cluster-scoped RBAC resources (clusterrole, clusterrolebinding), they will be able to escalate their privileges by giving their argocd more roles.

There's also the chance that if they can create privileged containers on the same cluster as the argocd install, they can "steal" service account tokens mounted by other pods and escalate that way, so make sure you have some way of preventing that from happening.

**Comment by jannfis** [comment]
Yes, indeed. It is important to understand that currently, people who can create an Application resource in Argo CD's control plane namespace (i.e. `argocd`), basically is thought of as an Argo CD administrator. Argo CD RBAC is (and can) only be enforced on the Argo CD API level.

Please see [this proposal](https://github.com/argoproj/argo-cd/pull/6409) for improvement on these mechanisms. If it would solve your use case, please participate in the discussion or give it a thumbs up. One of the major goals of that proposal is to allow people to create Applications in a declarative way (in their own namespaces), while Argo CD will be able to enforce project-level restrictions on those Applications.
